### ISSUE TITLE
Use iterative machine mixin in more algorithms

### ISSUE THREAD

**Issue by shubham808** [shogun-toolbox/shogun#4488]
This is a very good issue for those who want to be involved in the project [inside the black box II](https://github.com/shogun-toolbox/shogun/wiki/GSoC_2019_inside_the_black_box-II)

In order to make algorithms stoppable, we need to execute each iteration and then save the state in member variables. If a user decides to stop the process at any iteration, control is returned back. This means the user is free to continue, serialize, test etc. on the pre-trained model
See [Perceptron](https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/classifier/Perceptron.cpp#L35) and [NewtonSVM](https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/classifier/svm/NewtonSVM.cpp#L50) as examples.
We want all iterative algorithms to apply [this](https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/machine/IterativeMachine.h) mixin
#### First Steps
* Identify the initialization and iteration phase(what occurs in the while loop) of the algorithm
* The initialization part (mostly above the while loop) goes into the ```init_model()``` method
* The iteration function will implement a single iteration of the algorithm, so we will move the contents of the while loop into ```iteration``` function
* Create new member variables if required to share data between ```init_model``` and ```iteration```
* Delete local copies of variables that represent state(like weights and bias)
* Finally write tests to see if everything is consistent :) see [NewtonSVM unit tests](https://github.com/shogun-toolbox/shogun/blob/develop/tests/unit/classifier/NewtonSVM_unittest.cc) and [Perceptron unit tests](https://github.com/shogun-toolbox/shogun/blob/develop/tests/unit/classifier/Perceptron_unittest.cc#L92)

Initially we want this done for a single Linear Machine algorithm ([CAveragedPerceptron](https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/classifier/AveragedPerceptron.cpp) is a good place to start)
See the [iterative machine guide](https://shubham808.github.io/blog/gsoc'18/IterativeMachine) for more information

**Comment by souvik3333** [comment]
Hi @shubham808, I am new here. I read the wiki for inside the black box II project and I am really interested to work on it. This issue would be a great introduction to the project. So, Can I work on this?

**Comment by shubham808** [comment]
Yes go ahead :)

**Comment by han0305** [comment]
I am new to open source and I will like to work on this issue.

**Comment by samdbrice** [comment]
I this issue closed as per the Merge on Mar 2?

**Comment by karlnapf** [comment]
Nope, still lots of iterative algorithms/machines to port

**Comment by samdbrice** [comment]
Ah ok, so this is akin to a Milestone/Sticky issue - not a one-off item. This is indeed a good first issue, I'll see if I can implement one.

**Comment by karlnapf** [comment]
Just checked a few machines that are iterative and realised that most low hanging fruits have been taken. There are quite a few algorithms to port left but the changes will be slightly more nontrivial. Everything that is based on (stochastic) gradient descent is worth looking into here

**Comment by progmatic-99** [comment]
I would like to contribute. Is there any work left to be done in this issue?

**Comment by ashutosh-b-b** [comment]
Hi. If this is still open can I take up this one?

**Comment by sbrice** [comment]
Sure, go for it. Thanks!

**Comment by ashutosh-b-b** [comment]
@samdbrice I did started to read the blog and understand it. Which algorithm should I go with knowing that I am a beginner here?

**Comment by ashutosh-b-b** [comment]
I was thinking to start with Brute KNN can anyone guide me?

**Comment by vigsterkr** [comment]
@ashutosh-b-b you could start with that or with https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/metric/LMNN.cpp basically the idea is that you follow the CRTP pattern with IterativeMachine and move an iteration to logic into `iteration` function

**Comment by ashutosh-b-b** [comment]
Thanks!
